File: frontend/js/circuit/qasmParser.js
/* ============================================================
   qasmParser.js — OpenQASM 2.0 → Circuit Model
   ============================================================
   PURPOSE:
     Parses an OpenQASM 2.0 string and converts it back into
     an array of gate objects for the CircuitModel. This is
     the "Code → GUI" direction.

   RESPONSIBILITIES:
     - Parse the QASM header (version check)
     - Parse register declarations (qreg, creg)
     - Parse gate instructions into gate objects
     - Handle parameterized gates
     - Handle controlled gates
     - Handle SWAP, Measure, Reset
     - Assign column numbers (time-step ordering)
     - Report parsing errors

   DATA FLOW:
     OpenQASM 2.0 string
       → qasmParser.parse(qasmString)
       → Array of gate objects
       → model.setGates(gates)
       → Renderer re-renders

   HOW IT WORKS — STEP BY STEP:

     Input QASM:
       h q[0];
       cx q[0],q[7];

     Step 1: Tokenize each line
     Step 2: Match against known patterns
     Step 3: Create gate objects:
       { type: "H", targets: [0], column: 0 }
       { type: "CX", controls: [0], targets: [7], column: 1 }
     Step 4: Return the gates array
   ============================================================ */

/**
 * QASMParser — Converts OpenQASM 2.0 text to gate objects.
 */
export class QASMParser {
  constructor() {
    // ── Reverse mapping: QASM instruction → gate type ────
    this.instructionMap = {
      'h':       'H',
      'x':       'X',
      'y':       'Y',
      'z':       'Z',
      's':       'S',
      'sdg':     'SDG',
      't':       'T',
      'tdg':     'TDG',
      'p':       'P',
      'rx':      'RX',
      'ry':      'RY',
      'rz':      'RZ',
      'cx':      'CX',
      'cy':      'CY',
      'cz':      'CZ',
      'ccx':     'CCX',
      'swap':    'SWAP',
      'measure': 'MEASURE',
      'reset':   'RESET',
    };

    // ── Gate properties ──────────────────────────────────
    // How many qubits each gate uses, and which are controls
    this.gateProperties = {
      H:  { controls: 0, targets: 1 },
      X:  { controls: 0, targets: 1 },
      Y:  { controls: 0, targets: 1 },
      Z:  { controls: 0, targets: 1 },
      S:  { controls: 0, targets: 1 },
      SDG: { controls: 0, targets: 1 },
      T:  { controls: 0, targets: 1 },
      TDG: { controls: 0, targets: 1 },
      P:  { controls: 0, targets: 1, hasParams: true },
      RX: { controls: 0, targets: 1, hasParams: true },
      RY: { controls: 0, targets: 1, hasParams: true },
      RZ: { controls: 0, targets: 1, hasParams: true },
      CX: { controls: 1, targets: 1 },
      CY: { controls: 1, targets: 1 },
      CZ: { controls: 1, targets: 1 },
      CCX: { controls: 2, targets: 1 },
      SWAP: { controls: 0, targets: 2 },
      MEASURE: { controls: 0, targets: 1 },
      RESET: { controls: 0, targets: 1 },
    };
  }

  /**
   * Parse an OpenQASM 2.0 string into an array of gate objects.
   *
   * @param {string} qasmString — The QASM source code
   * @returns {{ gates: Array<Object>, errors: Array<string>, numQubits: number }}
   *
   * @example
   *   const result = parser.parse(`
   *     OPENQASM 2.0;
   *     include "qelib1.inc";
   *     qreg q[8];
   *     creg c[8];
   *     h q[0];
   *     cx q[0],q[7];
   *   `);
   *   // result.gates = [
   *   //   { type: "H", targets: [0], controls: [], column: 0, params: {} },
   *   //   { type: "CX", targets: [7], controls: [0], column: 1, params: {} },
   *   // ]
   */
  parse(qasmString) {
    const gates = [];
    const errors = [];
    let numQubits = 8;  // default
    let columnCounter = 0;

    // Track which qubits are used in each column to assign columns properly
    const columnUsage = [];  // columnUsage[col] = Set of qubit indices

    const lines = qasmString.split('\n');

    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
      let line = lines[lineNum].trim();

      // Skip empty lines and comments
      if (!line || line.startsWith('//') || line.startsWith('/*')) continue;

      // Remove trailing semicolons and whitespace
      line = line.replace(/;$/, '').trim();

      // ── Header lines ────────────────────────────────
      if (line.startsWith('OPENQASM')) continue;
      if (line.startsWith('include')) continue;

      // ── Register declarations ───────────────────────
      const qregMatch = line.match(/^qreg\s+\w+\[(\d+)\]/);
      if (qregMatch) {
        numQubits = parseInt(qregMatch[1]);
        continue;
      }
      if (line.startsWith('creg')) continue;

      // ── Barrier ─────────────────────────────────────
      if (line.startsWith('barrier')) continue;

      // ── Gate instructions ───────────────────────────
      const gateResult = this._parseGateInstruction(line, lineNum);
      if (gateResult.error) {
        errors.push(`Line ${lineNum + 1}: ${gateResult.error}`);
        continue;
      }
      if (gateResult.gate) {
        // Assign column: find the first column where none of the
        // involved qubits are already used
        const involvedQubits = [...gateResult.gate.controls, ...gateResult.gate.targets];
        let col = 0;
        while (true) {
          if (!columnUsage[col]) columnUsage[col] = new Set();
          const conflict = involvedQubits.some(q => columnUsage[col].has(q));
          if (!conflict) break;
          col++;
        }
        // Mark qubits as used in this column
        if (!columnUsage[col]) columnUsage[col] = new Set();
        involvedQubits.forEach(q => columnUsage[col].add(q));

        gateResult.gate.column = col;
        gates.push(gateResult.gate);
      }
    }

    return { gates, errors, numQubits };
  }

  /**
   * Parse a single gate instruction line.
   *
   * @param {string} line — The instruction (without semicolon)
   * @param {number} lineNum — Line number for error reporting
   * @returns {{ gate: Object|null, error: string|null }}
   *
   * SUPPORTED PATTERNS:
   *   "h q[0]"                → H on q0
   *   "rx(1.5708) q[0]"      → RX(1.5708) on q0
   *   "cx q[0],q[7]"         → CX: control q0, target q7
   *   "ccx q[0],q[1],q[2]"   → CCX: controls q0,q1, target q2
   *   "swap q[2],q[5]"       → SWAP q2, q5
   *   "measure q[0] -> c[0]" → Measure q0
   *   "reset q[3]"           → Reset q3
   */
  _parseGateInstruction(line, lineNum) {
    // ── Measure: "measure q[n] -> c[n]" ─────────────────
    const measureMatch = line.match(/^measure\s+q\[(\d+)\]\s*->\s*c\[(\d+)\]/);
    if (measureMatch) {
      return {
        gate: {
          type: 'MEASURE',
          controls: [],
          targets: [parseInt(measureMatch[1])],
          params: {},
        },
        error: null,
      };
    }

    // ── Reset: "reset q[n]" ─────────────────────────────
    const resetMatch = line.match(/^reset\s+q\[(\d+)\]/);
    if (resetMatch) {
      return {
        gate: {
          type: 'RESET',
          controls: [],
          targets: [parseInt(resetMatch[1])],
          params: {},
        },
        error: null,
      };
    }

    // ── Parameterized: "rx(theta) q[n]" ─────────────────
    const paramMatch = line.match(/^(\w+)\(([\d.e\-+]+)\)\s+(.+)/);
    if (paramMatch) {
      const instrName = paramMatch[1].toLowerCase();
      const theta = parseFloat(paramMatch[2]);
      const qubitsStr = paramMatch[3];

      const gateType = this.instructionMap[instrName];
      if (!gateType) {
        return { gate: null, error: `Unknown gate: ${instrName}` };
      }

      const qubits = this._parseQubitArgs(qubitsStr);
      if (!qubits) {
        return { gate: null, error: `Invalid qubit arguments: ${qubitsStr}` };
      }

      return {
        gate: {
          type: gateType,
          controls: [],
          targets: qubits,
          params: { theta },
        },
        error: null,
      };
    }

    // ── Standard: "gatename q[a],q[b],..." ──────────────
    const standardMatch = line.match(/^(\w+)\s+(.+)/);
    if (standardMatch) {
      const instrName = standardMatch[1].toLowerCase();
      const qubitsStr = standardMatch[2];

      const gateType = this.instructionMap[instrName];
      if (!gateType) {
        return { gate: null, error: `Unknown gate: ${instrName}` };
      }

      const qubits = this._parseQubitArgs(qubitsStr);
      if (!qubits) {
        return { gate: null, error: `Invalid qubit arguments: ${qubitsStr}` };
      }

      const props = this.gateProperties[gateType];
      if (!props) {
        return { gate: null, error: `Unknown gate properties: ${gateType}` };
      }

      // Split qubits into controls and targets
      const controls = qubits.slice(0, props.controls);
      const targets = qubits.slice(props.controls);

      return {
        gate: {
          type: gateType,
          controls,
          targets,
          params: {},
        },
        error: null,
      };
    }

    return { gate: null, error: `Could not parse: ${line}` };
  }

  /**
   * Parse qubit arguments like "q[0],q[7]" into [0, 7].
   *
   * @param {string} str — Qubit arguments string
   * @returns {number[]|null} — Array of qubit indices, or null on error
   */
  _parseQubitArgs(str) {
    const matches = str.match(/q\[(\d+)\]/g);
    if (!matches) return null;

    return matches.map(m => {
      const numMatch = m.match(/\[(\d+)\]/);
      return numMatch ? parseInt(numMatch[1]) : NaN;
    }).filter(n => !isNaN(n));
  }
}

File: backend/qasm_parser.py
"""
============================================================
qasm_parser.py — Server-Side QASM Parser
============================================================
PURPOSE:
    Parses an OpenQASM 2.0 string into a list of gate objects
    (circuit model). This mirrors the frontend's qasmParser.js
    but runs on the server for validation and verification.

RESPONSIBILITIES:
    - Parse QASM header and register declarations
    - Parse gate instructions into gate dictionaries
    - Handle parameterized, controlled, and special gates
    - Assign column numbers for visualization
    - Report parsing errors
============================================================
"""

import re
import json
from typing import List, Dict, Any, Tuple, Optional


# ── Instruction Name to Gate Type ────────────────────────────
INSTRUCTION_MAP = {
    "h":       "H",
    "x":       "X",
    "y":       "Y",
    "z":       "Z",
    "s":       "S",
    "sdg":     "SDG",
    "t":       "T",
    "tdg":     "TDG",
    "p":       "P",
    "rx":      "RX",
    "ry":      "RY",
    "rz":      "RZ",
    "cx":      "CX",
    "cy":      "CY",
    "cz":      "CZ",
    "ccx":     "CCX",
    "swap":    "SWAP",
    "measure": "MEASURE",
    "reset":   "RESET",
}

# ── Gate Properties (controls count) ─────────────────────────
GATE_CONTROLS = {
    "H": 0, "X": 0, "Y": 0, "Z": 0,
    "S": 0, "SDG": 0, "T": 0, "TDG": 0, "P": 0,
    "RX": 0, "RY": 0, "RZ": 0,
    "CX": 1, "CY": 1, "CZ": 1,
    "CCX": 2,
    "SWAP": 0, "MEASURE": 0, "RESET": 0,
}


def parse_qasm(qasm_string: str) -> Dict[str, Any]:
    """
    Parse OpenQASM 2.0 into a circuit model.
    
    Args:
        qasm_string: Valid OpenQASM 2.0 source code
    
    Returns:
        dict: {
            "gates": [
                {"type": "H", "targets": [0], "controls": [], "column": 0, "params": {}},
                ...
            ],
            "errors": ["Line 5: Unknown gate: foo"],
            "num_qubits": 8
        }
    """
    gates = []
    errors = []
    num_qubits = 8
    
    # Track which qubits are used in each column
    column_usage: List[set] = []
    
    lines = qasm_string.strip().split("\n")
    
    for line_num, raw_line in enumerate(lines, start=1):
        line = raw_line.strip()
        
        # Skip empty lines and comments
        if not line or line.startswith("//") or line.startswith("/*"):
            continue
        
        # Remove trailing semicolons
        line = line.rstrip(";").strip()
        
        # ── Header ──────────────────────────────────────
        if line.startswith("OPENQASM") or line.startswith("include"):
            continue
        
        # ── Register declarations ────────────────────────
        qreg_match = re.match(r"qreg\s+\w+\[(\d+)\]", line)
        if qreg_match:
            num_qubits = int(qreg_match.group(1))
            continue
        if line.startswith("creg"):
            continue
        if line.startswith("barrier"):
            continue
        
        # ── Parse gate instruction ───────────────────────
        gate, error = _parse_instruction(line)
        
        if error:
            errors.append(f"Line {line_num}: {error}")
            continue
        
        if gate:
            # Assign column
            involved = gate["controls"] + gate["targets"]
            col = _find_column(column_usage, involved)
            gate["column"] = col
            gates.append(gate)
    
    return {
        "gates": gates,
        "errors": errors,
        "num_qubits": num_qubits,
    }


def _parse_instruction(line: str) -> Tuple[Optional[Dict], Optional[str]]:
    """Parse a single QASM instruction line."""
    
    # ── Measure: "measure q[n] -> c[n]" ──────────────────
    measure_match = re.match(r"measure\s+q\[(\d+)\]\s*->\s*c\[(\d+)\]", line)
    if measure_match:
        return {
            "type": "MEASURE",
            "controls": [],
            "targets": [int(measure_match.group(1))],
            "params": {},
        }, None
    
    # ── Reset: "reset q[n]" ──────────────────────────────
    reset_match = re.match(r"reset\s+q\[(\d+)\]", line)
    if reset_match:
        return {
            "type": "RESET",
            "controls": [],
            "targets": [int(reset_match.group(1))],
            "params": {},
        }, None
    
    # ── Parameterized: "rx(theta) q[n]" ──────────────────
    param_match = re.match(r"(\w+)\(([\d.eE\-+]+)\)\s+(.+)", line)
    if param_match:
        instr_name = param_match.group(1).lower()
        theta = float(param_match.group(2))
        qubits_str = param_match.group(3)
        
        gate_type = INSTRUCTION_MAP.get(instr_name)
        if not gate_type:
            return None, f"Unknown gate: {instr_name}"
        
        qubits = _parse_qubits(qubits_str)
        if qubits is None:
            return None, f"Invalid qubit arguments: {qubits_str}"
        
        return {
            "type": gate_type,
            "controls": [],
            "targets": qubits,
            "params": {"theta": theta},
        }, None
    
    # ── Standard: "gate q[a],q[b],..." ───────────────────
    standard_match = re.match(r"(\w+)\s+(.+)", line)
    if standard_match:
        instr_name = standard_match.group(1).lower()
        qubits_str = standard_match.group(2)
        
        gate_type = INSTRUCTION_MAP.get(instr_name)
        if not gate_type:
            return None, f"Unknown gate: {instr_name}"
        
        qubits = _parse_qubits(qubits_str)
        if qubits is None:
            return None, f"Invalid qubit arguments: {qubits_str}"
        
        num_controls = GATE_CONTROLS.get(gate_type, 0)
        controls = qubits[:num_controls]
        targets = qubits[num_controls:]
        
        return {
            "type": gate_type,
            "controls": controls,
            "targets": targets,
            "params": {},
        }, None
    
    return None, f"Could not parse: {line}"


def _parse_qubits(qubits_str: str) -> Optional[List[int]]:
    """Parse qubit arguments like 'q[0],q[7]' into [0, 7]."""
    matches = re.findall(r"q\[(\d+)\]", qubits_str)
    if not matches:
        return None
    return [int(m) for m in matches]


def _find_column(column_usage: List[set], qubits: List[int]) -> int:
    """Find the first column where none of the qubits are used."""
    col = 0
    while True:
        if col >= len(column_usage):
            column_usage.append(set())
        if not any(q in column_usage[col] for q in qubits):
            for q in qubits:
                column_usage[col].add(q)
            return col
        col += 1
